packages/framework-core/src/services/read-model-store.ts

```diff
@@ -91,27 +70,206 @@ export class ReadModelStore {
+      if (!entityJoinKey) {
+        logger.warn(
+          `Couldn't find the joinKey named ${projectionMetadata} in entity snapshot of ${entityMetadata.class.name}. Skipping...`
+        )
+        return []
+      }
+      const entitiesJoinKeys: Array<UUID> = Array.isArray(entityJoinKey) ? entityJoinKey : [entityJoinKey]
+      // A new Read Model with JoinKey by entity needs to be projected using a JoinKey with an entity query. We don't have a previous read model, but we have the new id from the entity
+      for (const readModelId of entitiesJoinKeys) {
+        const readModel = currentReadModel?.id === readModelId ? currentReadModel : undefined
+        projections.push(
+          this.projectAndStoreReadModelWithRetry(
+            entitySnapshotEnvelope,
+            readModelName,
+            sequenceKey,
+            entityInstance,
+            projectionMetadata,
+            readModelId,
+            readModel
+          )
+        )
+      }
+    } else {
+      // A new Read Model with JoinKey by ReadModel needs to be projected using a JoinKey with a read model query. We don't have a previous read model nor the new id as we don't have an entity field
+      projections.push(
+        this.projectAndStoreReadModelWithRetry(
+          entitySnapshotEnvelope,
+          readModelName,
+          sequenceKey,
+          entityInstance,
+          projectionMetadata,
+          currentReadModel?.id,
+          currentReadModel
+        )
+      )
+    }
+    return projections
+  }
+
+  private async projectAndStoreReadModelWithRetry(
+    entitySnapshotEnvelope: EntitySnapshotEnvelope,
+    readModelName: string,
+    sequenceKey: SequenceKey | undefined,
+    entityInstance: EntityInterface,
+    projectionMetadata: ProjectionMetadata<EntityInterface, ReadModelInterface>,
+    readModelId?: UUID,
+    currentReadModel?: ReadModelInterface
+  ): Promise<unknown> {
+    const logger = getLogger(this.config, 'ReadModelStore#projectAndStoreReadModelWithRetry')
+    logger.debug(
+      'Projecting entity snapshot ',
+      entitySnapshotEnvelope,
+      ` to build new state of read model ${readModelName} with ID ${currentReadModel?.id || readModelId}`,
+      sequenceKey ? ` sequencing by ${sequenceKey.name} with value ${sequenceKey.value}` : ''
+    )
+
+    return retryIfError(
+      (tryNumber?: number) =>
+        this.applyProjectionToReadModel(
+          entityInstance,
+          projectionMetadata,
+          currentReadModel,
+          entitySnapshotEnvelope,
+          readModelId,
+          sequenceKey,
+          tryNumber
+        ),
+      OptimisticConcurrencyUnexpectedVersionError,
+      logger
+    )
+  }
+
+  public async getReadModels(
+    entityInstance: EntityInterface,
+    projectionMetadata: ProjectionMetadata<EntityInterface, ReadModelInterface>,
+    entityMetadata: EntityMetadata
+  ): Promise<Array<ReadModelInterface>> {
```

Comment:
- I guess this is going to be the new public method, replacing `fetchReadModel`, isn't it? The latter now is non-public because of that. Is this assumption correct?
- If this is going to be the new public method, could you add a comment on top of the method signature?

Reply:
Comment added in commit 1786d58d0167e3e16ecb19db614fac40ceaebebf.